src/LVMGenerator.h

```diff
@@ -104,6 +176,146 @@ class RelationEncoder {
+    }
+
+    /**
+     * Add nested operation which require a view (Hints).
+     * This node cannot come from the outer-most filter operation.
+     */
+    void addViewForNestedOperation(const RamNode* op) {
+        assert(dynamic_cast<const RamIndexOperation*>(op) || dynamic_cast<const RamExistenceCheck*>(op) ||
+                dynamic_cast<const RamProvenanceExistenceCheck*>(op));
+        viewsForNestedOperation.push_back(op);
+    }
+
+    /**
+     * Add a list of nested operations which require a view.
+     */
+    void addNestedOperationList(const std::vector<const RamNode*>& nodes) {
+        for (auto& node : nodes) {
+            addViewForNestedOperation(node);
+        }
+    }
+
+    /**
+     * Return outer-most filter operations.
+     */
+    const std::vector<const RamCondition*>& getOuterFilterOperation() {
+        return outerOperations;
+    }
+
+    /**
+     * Return views for outer-most filter operations.
+     */
+    const std::vector<const RamNode*>& getViewsInOuterOperation() {
+        return viewsForFilterOperation;
+    }
+
+    /**
+     * Return nested operations
+     */
+    std::vector<const RamNode*> getViewsInNestedOperation() {
+        return viewsForNestedOperation;
+    }
+
+    /**
+     * Find all operations under the root node that require a view.
+     *
+     * In particular, find all index operation and existence check operation.
+     */
+    std::vector<const RamNode*> findAllViews(const RamNode& node) {
+        std::vector<const RamNode*> res;
+        visitDepthFirst(node, [&](const RamNode& node) {
+            if (requireView(&node) == true) {
+                res.push_back(&node);
+            };
+        });
+        return res;
+    }
+
+    /**
+     * Return true if the given operation requires a view.
+     */
+    bool requireView(const RamNode* node) {
+        if (auto exists = dynamic_cast<const RamExistenceCheck*>(node)) {
+            for (auto& v : exists->getValues()) {
+                if (!isRamUndefValue(v)) {
+                    return true;
+                }
+            }
+            return false;
+        } else if (auto provExists = dynamic_cast<const RamProvenanceExistenceCheck*>(node)) {
+            for (auto& v : provExists->getValues()) {
+                if (!isRamUndefValue(v)) {
+                    return true;
+                }
+            }
+            return false;
+        } else if (auto indexOp = dynamic_cast<const RamIndexOperation*>(node)) {
+            for (auto& v : indexOp->getRangePattern()) {
+                if (!isRamUndefValue(v)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+        return false;
+    }
+
+    void reset() {
+        outerOperations.clear();
+        viewsForFilterOperation.clear();
+        viewsForNestedOperation.clear();
+    }
+
+    /**
+     * Same as toConjunctionList defined in RamOperation. But does not clone new node,
```

Comment:
I think you mean in RamCondition.
It's worth repeating the description, so the reader doesn't have to look in the other file for the description.